design/20241124-rename-bunde-to-clusterbundle.md

```diff
@@ -0,0 +1,162 @@
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+TODO: add more items to checklist
+
+## Summary
+
+We want to rename `Bundle` to `ClusterBundle` for reasons described in "Motivation" below.
+
+Conversion webhooks in Kubernetes can only be used to convert between versions of an API,
+and not to convert between resources. This makes renaming harder to do, as it will represent
+deleting the old API and creating a new API.
+
+Even though trust-manager `Bundle` API is officially in an alpha state, we want to provide a smooth
+migration path for our users. For this reason we are proposing to do this change over multiple releases,
+requiring end-user action on the way, but without risking trust-manager target resources - which can provide
+critical features in a cluster. For additional details, please see "Proposal" below.
+
+## Motivation
+
+We have [an open issue](https://github.com/cert-manager/trust-manager/issues/63) suggesting renaming Bundle to ClusterBundle.
+Might seem like an obvious WontFix, considering all the assumed work (and pain) to do this,
+but it turns out that the current name somehow blocks introducing an equivalent namespace-scoped resource -
+for trust bundle management inside namespaces in multi-tenant clusters.
+
+In addition, `ClusterBundle` would be a better name looking at cert-manager having `Issuer` and `ClusterIssuer`.
+The name would also be better aligned with upstream [ClusterTrustBundle](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#cluster-trust-bundles).
+
+### Goals
+
+- `Bundle` resource is renamed to `ClusterBundle`
+- Provide a smooth migration for users of trust-manager
+- Use the opportunity window to improve the API
+  - [Issue summarizing proposed API changes (#242)](https://github.com/cert-manager/trust-manager/issues/242)
+  - [Draft PR for proposed API changes done when migrating (#486)](https://github.com/cert-manager/trust-manager/pull/486)
+
+
+### Non-Goals
+
+- Behavior changes in `Bundle`/`ClusterBundle` API while migrating (only simpler API improvements)
+- Introduce a new API version
+- Introduce a namespace-scoped bundle API (at least not now)
+
+## Proposal
+
+We propose to do this change over multiple releases:
+
+1. Version X: Only `Bundle` API is served and has the functional controller. (present state)
+2. Version X+1: We introduce `ClusterBundle` with API otherwise identical to `Bundle`:
+   - `Bundle` is marked as deprecated.
+   - `ClusterBundle` introduced with agreed API improvements.
+   - Functional controller is migrated to `ClusterBundle`
+   - Validating webhook for both `Bundle` and `ClusterBundle` with same validations
+   - New technical controller added for `Bundle`, converting bundles to cluster bundles
+3. Version Y: Remove `Bundle` and only serve `ClusterBundle` (desired state)
```

Comment:
question: Is there any reason we couldn't have the following:

1. Version X: Status quo
2. Version X+1: Introduce `ClusterBundle` resource
    - Totally separate controller to `Bundle`
    - `Bundle` completely unaffected
    - Same `BundleSpec` shared between both resources
    - Warn in logs whenever `Bundle` is created (saying users should be creating `ClusterBundle`s instead)
3. Version Y: Start migrating `Bundle` resources to `ClusterBundle`s
4. Version Z: Remove `Bundle`

We can start the migration at the same time as we add the new CRD, but it seems like this way we can make a bit of progress (i.e. add ClusterBundle) without impacting anything else?

Reply:
Since a `Bundle` and `ClusterBundle` could then address the same target resources (by name), we would have to this into account in controllers. In addition, we would have a LOT more code to maintain and test. This is something I don't recommend, but I will add this to the list of alternatives. 👍

Reply:
I have added your suggestion as an alternative now.